README.md
# sticky_web_scroll

Use this package if you need more customization on your web scroll displays.

See the [install instructions](https://pub.dev/packages/sticky_web_scroll/install).

## Quick reference

Since customization requires a lot of properties, here is a quick cheatsheet:

Property | What does it do
-------- | ---------------
landing    | Takes the 5% of the screen height(Customizable), collasp when scrolled vertically.
body  | The body of the main content.
landingHeight  | The height of the landing to stickyAppbar point.
sideNav  | Side widget(optional).
stickyAppBar  | Immediate below landing,(Customizable) stick to top when landing collasp .
stickyHeight  | stickyAppBar, stickyAppBar height.(stickyHeight must not be greater than height of landingHeight)

## We are on YouTube!

You can see a video tutorial here.
1.Awesome Flutter Tutorials [tutorial](https://www.youtube.com/channel/UC1fuQBjVLSqwk-PgdeeMAyQ) by **Flutter Developer's Core**

To use this package :

* add the dependency to your [pubspec.yaml](https://github.com/hello-paulvin/sticky-web-scroll/pubspec.yaml) file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    sticky_web_scroll: 
```

## Getting Started

The examples below were updated for version 0.0.1. Changes might have been made. See the [changelog](CHANGELOG.md).

## Screenshot

![sticky_web_scroll Animated](screenshots/screen1.gif)

### A basic Sticky-Web-Scroll 

```dart

class HomePage extends StatelessWidget {
  const HomePage({
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: StickyWebScroll(
        landingHeight: 400,
        stickyHeight: 80,
        landing: Container(
          decoration: BoxDecoration(
            image: DecorationImage(
                image: NetworkImage(
                    'https://images.unsplash.com/photo-1489843704329-b420d20a0b43?ixid=MXwxMjA3fDB8MHxjb2xsZWN0aW9uLXBhZ2V8MXwyMzA0MDE1fHxlbnwwfHx8&ixlib=rb-1.2.1&w=1000&q=80'),
                fit: BoxFit.cover),
          ),
        ),
        stickyAppBar: Container(
          color: Colors.blue,
          width: MediaQuery.of(context).size.width,
          child: Center(
            child: Text('Stick Web Scroll - by Flutter Devs Core',
                style: Theme.of(context).textTheme.headline4),
          ),
        ),
        body: SingleChildScrollView(
          physics: NeverScrollableScrollPhysics(),
          child: Container(
            color: Colors.white,
            width: MediaQuery.of(context).size.width,
            child: Column(
              children: List.generate(
                100,
                (index) => Container(
                  height: 500,
                  width: 500,
                  color: Colors
                      .primaries[Random().nextInt(Colors.primaries.length)],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## 0.0.1 - 2021-03-09
- First Release
## 0.0.2 - 2021-03-12
- Bug fix Second Release


test/sticky_web_scroll_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:sticky_web_scroll/sticky_web_scroll.dart';

void main() {
  testWidgets('sticky-web-scroll-test', (WidgetTester tester) async {
    final stickyScroll = StickyWebScroll(
      landingHeight: 200,
      stickyHeight: 80,
      landing: SizedBox(),
      stickyAppBar: SizedBox(),
      body: SizedBox(),
      pinned: true,
      floating: true,
      restorationId: 'test',
      scrollDirection: Axis.vertical,
      clipBehavior: Clip.hardEdge,
      reverse: false,
    );

    expect(stickyScroll, equals(stickyScroll));
  });
}


lib/sticky_landing.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class StickyLanding extends SliverPersistentHeaderDelegate {
  final double collapseHeight;
  final double bottomHeight;
  final FloatingHeaderSnapConfiguration snap;
  final Widget collapse;
  final Widget bottom;

  StickyLanding({
    @required this.bottomHeight,
    @required this.collapse,
    @required this.bottom,
    @required this.collapseHeight,
    this.snap,
  });

  @override
  Widget build(
      BuildContext context, double shrinkOffset, bool overlapsContent) {
    final firstBarOffset = (-shrinkOffset * 0.01) - shrinkOffset;
    final secondBarOffset = (-shrinkOffset * 1) + shrinkOffset;

    return Container(
      color: Colors.white,
      child: Stack(
        fit: StackFit.expand,
        children: [
          Positioned(
            bottom: secondBarOffset,
            child: Container(
              width: MediaQuery.of(context).size.width,
              child: Wrap(
                children: [
                  Container(
                    height: bottomHeight + 3,
                    child: bottom ?? SizedBox(),
                  )
                ],
              ),
            ),
          ),
          Positioned(
            top: firstBarOffset - bottomHeight,
            child: Container(
              width: MediaQuery.of(context).size.width,
              child: Wrap(
                children: [
                  Container(
                    height: collapseHeight,
                    child: collapse ?? SizedBox(),
                  )
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }

  @override
  double get maxExtent => collapseHeight;

  @override
  double get minExtent => bottomHeight;

  @override
  FloatingHeaderSnapConfiguration get snapConfiguration => snap;

  @override
  bool shouldRebuild(SliverPersistentHeaderDelegate oldDelegate) => true;
}


lib/sticky_web_scroll.dart
library sticky_web_scroll;

import 'package:flutter/material.dart';
import 'package:sticky_web_scroll/sticky_landing.dart';

enum SideNavPosition { Left, Right }

class StickyWebScroll extends StatelessWidget {
  final Widget landing;
  final Widget body;
  final double landingHeight;
  final Widget sideNav;
  final Widget stickyAppBar;
  final double stickyHeight;
  final bool floating;
  final bool pinned;
  final ScrollController controller;
  final Axis scrollDirection;
  final ScrollPhysics physics;
  final bool reverse;
  final Clip clipBehavior;
  final String restorationId;
  final SideNavPosition sideNavPosition;

  const StickyWebScroll({
    Key key,
    @required this.body,
    @required this.landing,
    @required this.stickyAppBar,
    this.sideNav,
    this.landingHeight,
    this.stickyHeight,
    this.floating = true,
    this.pinned = true,
    this.controller,
    this.clipBehavior = Clip.hardEdge,
    this.restorationId,
    this.scrollDirection = Axis.vertical,
    this.physics,
    this.reverse = false,
    this.sideNavPosition,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final screenSize = MediaQuery.of(context).size;
    final initialHeight = screenSize.height * 0.5;
    return NestedScrollView(
      controller: controller,
      scrollDirection: scrollDirection,
      physics: physics,
      reverse: reverse,
      clipBehavior: clipBehavior,
      restorationId: restorationId,
      body: Row(
        children: [
          sideNavPosition == null
              ? SizedBox()
              : sideNavPosition == SideNavPosition.Left
                  ? _sideNav()
                  : SizedBox(),
          _buildBody(),
          sideNavPosition == null
              ? SizedBox()
              : sideNavPosition == SideNavPosition.Right
                  ? _sideNav()
                  : SizedBox()
        ],
      ),
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return <Widget>[
          SliverOverlapAbsorber(
            handle: NestedScrollView.sliverOverlapAbsorberHandleFor(context),
            sliver: SliverPersistentHeader(
              floating: floating,
              pinned: pinned,
              delegate: StickyLanding(
                collapse: landing,
                collapseHeight: landingHeight ?? initialHeight,
                bottom: stickyAppBar,
                bottomHeight: stickyHeight ?? 50,
              ),
            ),
          ),
        ];
      },
    );
  }

  Widget _sideNav() => Wrap(children: [sideNav ?? Container()]);
  Widget _buildBody() => body;
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:sticky_web_scroll/sticky_web_scroll.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: StickyWebScroll(
        landingHeight: 400,
        stickyHeight: 80,
        landing: Container(
          decoration: BoxDecoration(
            image: DecorationImage(
                image: NetworkImage(
                    'https://images.unsplash.com/photo-1489843704329-b420d20a0b43?ixid=MXwxMjA3fDB8MHxjb2xsZWN0aW9uLXBhZ2V8MXwyMzA0MDE1fHxlbnwwfHx8&ixlib=rb-1.2.1&w=1000&q=80'),
                fit: BoxFit.cover),
          ),
        ),
        stickyAppBar: Container(
          color: Colors.blue,
          width: MediaQuery.of(context).size.width,
          child: Center(
            child: Text('Stick Web Scroll - by Flutter Devs Core',
                style: Theme.of(context).textTheme.headline4),
          ),
        ),
        body: SingleChildScrollView(
          physics: NeverScrollableScrollPhysics(),
          child: Container(
            color: Colors.white,
            width: MediaQuery.of(context).size.width,
            child: Column(
              children: List.generate(
                100,
                (index) => Container(
                  height: 500,
                  width: 500,
                  color: Colors
                      .primaries[Random().nextInt(Colors.primaries.length)],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
